content-based-message-routing/main.bal

```diff
@@ -0,0 +1,88 @@
+import ballerina/http;
+import ballerina/log;
+
+configurable int port = 8290;
+
+final http:Client grandOaksEP = check initializeHttpClient("http://localhost:9090/grandoaks/categories");
+final http:Client clemencyEP = check initializeHttpClient("http://localhost:9090/clemency/categories");
+final http:Client pineValleyEP = check initializeHttpClient("http://localhost:9090/pinevalley/categories");
+
+function initializeHttpClient(string url) returns http:Client|error => new (url);
+
+type Patient record {|
+    string name;
+    string dob;
+    string ssn;
+    string address;
+    string phone;
+    string email;
+|};
+
+type ReservationRequest record {|
+    Patient patient;
+    string doctor;
+    string hospital_id;
+    string hospital;
+    string appointment_date;
+|};
+
+type Doctor record {|
+    string name;
+    string hospital;
+    string category;
+    string availability;
+    float fee;
+|};
+
+type ReservationResponse record {|
+    int appointmentNumber;
+    Doctor doctor;
+    Patient patient;
+    float fee;
+    string hospital;
+    boolean confirmed;
+    string appointmentDate;
+|};
+
+service /healthcare on new http:Listener(port) {
+    resource function post categories/[string category]/reserve(ReservationRequest payload)
+            returns ReservationResponse|http:NotFound|http:BadRequest|http:InternalServerError {
+        http:Client? hospitalEP = ();
+        ReservationRequest {hospital_id, patient, ...reservationRequest} = payload;
+        match hospital_id {
```

Comment:
Alternatively, we can use an enum for hospital_id. That way we would ideally be able to do 

```ballerina
GRAND_OAKS => {
    log:printInfo("Routed to Grand Oak Community Hospital");
    hospitalEP = grandOaksEP;
}
CLEMENCY => {
    log:printInfo("Routed to Clemency Medical Center");
    hospitalEP = clemencyEP;
}
PINE_VALLEY => {
    log:printInfo("Routed to Pine Valley Community Hospital");
    hospitalEP = pineValleyEP;
}
```

Due to a bug, we still get the may be uninitialized error. But since we know that it can only be one of the three I'm OK with doing

```ballerina
GRAND_OAKS => {
    log:printInfo("Routed to Grand Oak Community Hospital");
    hospitalEP = grandOaksEP;
}
CLEMENCY => {
    log:printInfo("Routed to Clemency Medical Center");
    hospitalEP = clemencyEP;
}
_ => {
    log:printInfo("Routed to Pine Valley Community Hospital");
    hospitalEP = pineValleyEP;
}
```

Reply:
But then the logic is unnecessary when giving an error for `Routed to none`

Reply:
But why do we need to define an enum and implement it like this? We only have single usages for each hospital-id

Reply:
> But then the logic become unnecessary when giving an error for `Routed to none`

We won't have this then. Can basically get rid of it.



> But why do we need to define an enum and implement it like this? We only have single usages for each hospital-id

By defining an enum and using it in the record, the validation will happen at data-binding itself. That's the main thing, clean matching is like a bonus then.

Reply:
Any update on this?

Reply:
Fixed with https://github.com/ballerina-guides/integration-tutorials/pull/4/commits/12d8aca27c2be3839d51ceec80d1b39e350a6bd5